# Lecture 21. Writing to Tables (Hands On)

In this notebook, we are going to explore SQL syntax to insert and update records in Delta tables.

And remember Delta technology provides ACID compliant updates to Delta tables.

For this demonstration, we will continue working with our bookstore dataset.

<div style="text-align: center;">
<img src="../../assets/images/Presentation-Images/bookstore_schema.png" style="width:640px" >
</div> 

Let us first run our helping notebook to copy the dataset.

In [0]:
%run ../Includes/Copy-Datasets


And then we will use a CTAS statement to create `orders` delta table `AS SELECT` statement from parquet files.

In [0]:
%sql
CREATE TABLE orders AS
SELECT * FROM parquet.`${dataset.bookstore}/orders`

num_affected_rows num_inserted_rows

Let's query this table.

In [0]:
%sql
SELECT * FROM orders

order_id order_timestamp customer_id quantity total books 000000000003559 1657722056 C00001 2 48 List(List(B09, 2, 48)) 000000000004243 1658786901 C00002 2 55 List(List(B07, 1, 33), List(B06, 1, 22)) 000000000004321 1658934252 C00003 2 40 List(List(B04, 2, 40)) 000000000004392 1659034513 C00004 2 82 List(List(B08, 2, 82)) 000000000003673 1657934721 C00005 2 87 List(List(B01, 1, 49), List(B11, 1, 38)) 000000000004464 1659171834 C00006 2 77 List(List(B01, 1, 49), List(B02, 1, 28)) 000000000003495 1657541649 C00007 2 52 List(List(B09, 1, 24), List(B02, 1, 28)) 000000000004105 1658590490 C00008 2 69 List(List(B08, 1, 41), List(B02, 1, 28)) 000000000003825 1658166059 C00009 2 48 List(List(B09, 2, 48)) 000000000004062 1658519180 C00010 2 48 List(List(B09, 2, 48)) 000000000003650 1657906809 C00011 3 95 List(List(B08, 1, 41), List(B09, 1, 24), List(B12, 1, 30)) 000000000003951 1658343094 C00012 2 48 List(List(B09, 2, 48)) 000000000003910 1658275825 C00013 2 48 List(List(B09, 2, 48)) 000000000003525 1657627181 C00014 2 77 List(List(B07, 1, 33), List(B10, 1, 44)) 000000000004370 1659011452 C00015 2 85 List(List(B08, 1, 41), List(B10, 1, 44)) 000000000003500 1657554088 C00016 2 52 List(List(B09, 1, 24), List(B02, 1, 28)) 000000000004202 1658729125 C00017 3 104 List(List(B07, 1, 33), List(B12, 1, 30), List(B08, 1, 41)) 000000000003848 1658183282 C00018 2 48 List(List(B09, 2, 48)) 000000000003560 1657723619 C00019 3 95 List(List(B07, 1, 33), List(B11, 1, 38), List(B09, 1, 24)) 000000000003491 1657512356 C00020 2 40 List(List(B04, 2, 40)) 000000000003702 1657991039 C00021 3 105 List(List(B07, 1, 33), List(B02, 1, 28), List(B10, 1, 44)) 000000000003587 1657762562 C00022 4 122 List(List(B07, 1, 33), List(B06, 1, 22), List(B05, 1, 47), List(B04, 1, 20)) 000000000003672 1657934033 C00023 2 85 List(List(B08, 1, 41), List(B10, 1, 44)) 000000000003754 1658066470 C00024 2 68 List(List(B09, 1, 24), List(B10, 1, 44)) 000000000004039 1658494669 C00025 3 87 List(List(B08, 1, 41), List(B06, 1, 22), List(B09, 1, 24)) 000000000003744 1658058341 C00026 2 71 List(List(B07, 1, 33), List(B11, 1, 38)) 000000000003888 1658246267 C00027 2 48 List(List(B09, 2, 48)) 000000000004059 1658518516 C00028 2 66 List(List(B07, 2, 66)) 000000000004128 1658613154 C00029 2 77 List(List(B02, 1, 28), List(B01, 1, 49)) 000000000004204 1658734084 C00030 2 85 List(List(B08, 1, 41), List(B10, 1, 44)) 000000000003643 1657900371 C00031 3 126 List(List(B07, 1, 33), List(B01, 1, 49), List(B10, 1, 44)) 000000000003758 1658070984 C00032 2 55 List(List(B07, 1, 33), List(B06, 1, 22)) 000000000003577 1657744497 C00033 2 63 List(List(B07, 1, 33), List(B12, 1, 30)) 000000000003998 1658419690 C00034 2 69 List(List(B08, 1, 41), List(B02, 1, 28)) 000000000004124 1658608327 C00035 2 62 List(List(B09, 1, 24), List(B11, 1, 38)) 000000000004463 1659171199 C00036 2 68 List(List(B09, 1, 24), List(B10, 1, 44)) 000000000003693 1657985398 C00037 2 48 List(List(B09, 2, 48)) 000000000003520 1657585564 C00038 2 56 List(List(B02, 2, 56)) 000000000003796 1658141313 C00039 2 77 List(List(B07, 1, 33), List(B10, 1, 44)) 000000000003890 1658247999 C00040 2 77 List(List(B07, 1, 33), List(B10, 1, 44)) 000000000003600 1657813738 C00041 2 52 List(List(B02, 1, 28), List(B09, 1, 24)) 000000000004286 1658858563 C00042 2 79 List(List(B08, 1, 41), List(B11, 1, 38)) 000000000003598 1657810926 C00043 2 57 List(List(B03, 1, 35), List(B06, 1, 22)) 000000000004363 1659004174 C00044 2 48 List(List(B09, 2, 48)) 000000000003993 1658413622 C00045 3 87 List(List(B08, 1, 41), List(B06, 1, 22), List(B09, 1, 24)) 000000000003840 1658173729 C00046 2 71 List(List(B01, 1, 49), List(B06, 1, 22)) 000000000004475 1659191442 C00047 2 57 List(List(B09, 1, 24), List(B07, 1, 33)) 000000000004206 1658736598 C00048 2 48 List(List(B09, 2, 48)) 000000000004037 1658492334 C00049 2 46 List(List(B09, 1, 24), List(B06, 1, 22)) 000000000004032 1658482296 C00050 2 48 List(List(B09, 2, 48)) 000000000004458 1659152871 C00051 2 93 List(List(B01, 1, 49), Li

As you can see, parquet files have a well-defined schema, so we managed to extract the data correctly.

## Overwriting Tables

When writing to tables, we could be interested by completely overwriting the data in the table.

In fact, there are multiple **benefits to overwriting tables** instead of deleting and recreating tables.

  * For example, the old version of the table still exists and can easily retrieve all data using Time Travel.

  * In addition, overwriting a table is much faster because it does not need to list the directory recursively or delete any files.

  * In addition, it's an atomic operation.

Concurrent queries can still read the table while you are overwriting it.

And of course due to the ACID transaction guarantees, if overwriting the table fails, the table will be in its previous state.

### CRAS statement

The first method to accomplish complete overwrite is to use `CREATE OR REPLACE TABLE`. Also known as ***CRAS statement***.

`CREATE OR REPLACE TABLE` statements fully replace the content of a table each time they execute.

In [0]:
%sql
CREATE OR REPLACE TABLE orders AS
SELECT * FROM parquet.`${dataset.bookstore}/orders`

num_affected_rows num_inserted_rows

Let us now check our table history.

In [0]:
%sql
DESCRIBE HISTORY orders

version timestamp userId userName operation operationParameters job notebook clusterId readVersion isolationLevel isBlindAppend operationMetrics userMetadata engineInfo 1 2024-10-14T04:45:41Z 2895352578531874 suryapulika38@gmail.com CREATE OR REPLACE TABLE AS SELECT Map(partitionBy -> [], description -> null, isManaged -> true, properties -> {}, statsOnLoad -> false) null List(3802355948382795) 1011-150700-u18wk0fi 0 WriteSerializable false Map(numFiles -> 3, numOutputRows -> 2150, numOutputBytes -> 50780) null Databricks-Runtime/13.3.x-scala2.12 0 2024-10-14T04:42:39Z 2895352578531874 suryapulika38@gmail.com CREATE TABLE AS SELECT Map(partitionBy -> [], description -> null, isManaged -> true, properties -> {}, statsOnLoad -> false) null List(3802355948382795) 1011-150700-u18wk0fi null WriteSerializable true Map(numFiles -> 3, numOutputRows -> 2150, numOutputBytes -> 50780) null Databricks-Runtime/13.3.x-scala2.12

As you can see, the version 0 is a CREATE TABLE AS SELECT statement.

While, `CREATE OR REPLACE` statement has generated a new table version.

### `INSERT OVERWRITE` statement

The second method to overwrite table data is to use INSERT OVERWRITE statement.

It provides a nearly identical output as above.

It means data in the target table will be replaced by data from the query.

However, `INSERT OVERWRITE` statement has some differences.

- For example, it can only overwrite an existing table and not creating a new one like our CREATE OR REPLACE statement.

- And it can override only the new records that match the current table schema, which means that it is a safer technique for overwriting an existing table without the risk of modifying the table schema.

Let us run this command.

In [0]:
%sql
INSERT OVERWRITE orders
SELECT * FROM parquet.`${dataset.bookstore}/orders`

num_affected_rows num_inserted_rows 2150 2150

As you can see here, we have successfully overwriting the table data and rewriting 2150 records.

And we can see our table history again.

In [0]:
%sql
DESCRIBE HISTORY orders

version timestamp userId userName operation operationParameters job notebook clusterId readVersion isolationLevel isBlindAppend operationMetrics userMetadata engineInfo 2 2024-10-14T04:48:31Z 2895352578531874 suryapulika38@gmail.com WRITE Map(mode -> Overwrite, statsOnLoad -> false, partitionBy -> []) null List(3802355948382795) 1011-150700-u18wk0fi 1 WriteSerializable false Map(numFiles -> 3, numOutputRows -> 2150, numOutputBytes -> 50780) null Databricks-Runtime/13.3.x-scala2.12 1 2024-10-14T04:45:41Z 2895352578531874 suryapulika38@gmail.com CREATE OR REPLACE TABLE AS SELECT Map(partitionBy -> [], description -> null, isManaged -> true, properties -> {}, statsOnLoad -> false) null List(3802355948382795) 1011-150700-u18wk0fi 0 WriteSerializable false Map(numFiles -> 3, numOutputRows -> 2150, numOutputBytes -> 50780) null Databricks-Runtime/13.3.x-scala2.12 0 2024-10-14T04:42:39Z 2895352578531874 suryapulika38@gmail.com CREATE TABLE AS SELECT Map(partitionBy -> [], description -> null, isManaged -> true, properties -> {}, statsOnLoad -> false) null List(3802355948382795) 1011-150700-u18wk0fi null WriteSerializable true Map(numFiles -> 3, numOutputRows -> 2150, numOutputBytes -> 50780) null Databricks-Runtime/13.3.x-scala2.12


As you can see here, the `INSERT OVERWRITE` operation has been recorded as a new version in the table as `WRITE` operation.

And it has the modw "Overwrite".

And if you try to insert overwrite the data with different schema, for example, here we are adding a new column of the data for the current timestamp.

By running this command, we see that it generates an exception.

In [0]:
%sql
INSERT OVERWRITE orders
SELECT *, current_timestamp() FROM parquet.`${dataset.bookstore}/orders`

And the exception says "A schema mismatch detected when writing to the Delta table...".

So the way how they enforce schema on-write is the primary difference between `INSERT OVERWRITE` and `CREATE OR REPLACE TABLE` statements.

## Appending Data

Let us now talk about appending records to tables.

The easiest method is to use `INSERT INTO` statement.

Here we are inserting new data using an input query that query the parquet files in the `orders-new` directory.

In [0]:
%sql
INSERT INTO orders
SELECT * FROM parquet.`${dataset.bookstore}/orders-new`

num_affected_rows num_inserted_rows 700 700

We have successfully added 700 new records to our table.

And we can check the new number of orders.

In [0]:
%sql
SELECT count(*) FROM orders

count(1) 2850

Now we have 2850 records in the orders table.

## Merging Data

The `INSERT INTO` statement is a simple and efficient operation for inserting new data. 
However, it does not have any built in guarantees to prevent inserting the same records multiple times.
It means re-executing the query will write the same records to the target table resulting in duplicate records.

To resolve this issue, we can use our second method, which is `MERGE INTO` statement.

With the merge statement, you can upsert data from a source table, view, or dataframe into the target data table.
It means you can insert, update and delete using the `MERGE INTO` statements.

### An Example

Here, we will use the merge operation to update the customer data with updated emails and adding new customers.

- We are creating a temporary view of the new customer data.

- And now we can apply the merge operation that says `MERGE INTO customers`

  * the new changes coming from `customer_updates` temp view on the customer ID key.

  * And we have two actions here.

    When match, we do an update and when not match, we do an insert.

  * In addition, we add extra conditions.

    In this case, we are checking that the current row has a null email while the new record does not.

    In such a case, we update the email and we also update the last updated timestamp.

  * And again, if the new record does not match any existing customers based on the customer ID, in this case, we will insert this new record.

In [0]:
%sql
CREATE OR REPLACE TEMP VIEW customers_updates AS 
SELECT * FROM json.`${dataset.bookstore}/customers-json-new`;

MERGE INTO customers c
USING customers_updates u
ON c.customer_id = u.customer_id
WHEN MATCHED AND c.email IS NULL AND u.email IS NOT NULL THEN
  UPDATE SET email = u.email, updated = u.updated
WHEN NOT MATCHED THEN INSERT *

num_affected_rows num_updated_rows num_deleted_rows num_inserted_rows 301 100 0 201

As we can see here, we have updated 100 records and we have inserted 201 records.
And no records have been deleted.

So in a merge operation, updates, inserts and deletes are completed in a single atomic transaction.
In addition, merge operation is a great solution for avoiding duplicates when inserting records.

### Another Eexample

Here we have new books to be inserted and they are coming in CSV format.

We will create this temporary view against this new data.

In [0]:
%sql
CREATE OR REPLACE TEMP VIEW books_updates
   (book_id STRING, title STRING, author STRING, category STRING, price DOUBLE)
USING CSV
OPTIONS (
  path = "${dataset.bookstore}/books-csv-new",
  header = "true",
  delimiter = ";"
);

SELECT * FROM books_updates

book_id title author category price B14 Data Communications and Networking Behrouz A. Forouzan Computer Science 34.0 B15 Inside the Java Virtual Machine Bill Venners Computer Science 41.0 B13 Linux pocket guide Daniel J. Barrett Computer Science 26.0 B16 Green for Life Victoria Boutenko Food 18.0 B17 Cooking with Love Carla Hall Food 23.0

Here we have five new books and we are only interested by inserting the computer science books in our database.

Let us now use the `MERGE INTO` statement to update the table `books` with the data coming from the temporary view `books_updates`.

And now we can use the `MERGE INTO` statement where we provide only the not match condition.

It means we are only inserting new data if they are not already exist based on our key, which is the `book_id` and `the title`.

In addition, we are specifying the category of the new record to be inserted is only computer science.

In [0]:
%sql
MERGE INTO books b
USING books_updates u
ON b.book_id = u.book_id AND b.title = u.title
WHEN NOT MATCHED AND u.category = 'Computer Science' THEN 
  INSERT *

num_affected_rows num_updated_rows num_deleted_rows num_inserted_rows 3 0 0 3

As expected, we are only inserting three new records, which are the three computer science books.

And as we said, one of the main benefits of the merge operation is to avoid duplicates.

So if we try to rerun this statement, it will not reinsert those records as they are already on the table.

In [0]:
%sql
MERGE INTO books b
USING books_updates u
ON b.book_id = u.book_id AND b.title = u.title
WHEN NOT MATCHED AND u.category = 'Computer Science' THEN 
  INSERT *

num_affected_rows num_updated_rows num_deleted_rows num_inserted_rows 0 0 0 0

Yes, indeed. 
Zero record has been inserted.